### PR TITLE
[Memory Engine] Add TabletType to PartitionInfo and TabletMeta

### LINKED ISSUE #3442: [Memory Engine] Add MemTablet type to TabletMeta
To add MemTablet type to meta, there are 2 ways:

1. Create new meta kind, like BaseTablet and MemTablet
2. Use original meta, add new fields.

Unlike original Tablet, MemTablet doesn't have components like rowsets and may not have del_predicates, alter tasks etc. Those are vectors/ptrs so it won't use much memory(or pb space) if they are empty. Also it's worth to mention now MemTablet can handle multiple schema(schema change) inside MemTablet, it's won't be necessary to create a new tablet after schema change, but we can also keep the old behavior, depending on which way is easier.

To make things simple, we can first add a field, e.g. `bool is_memory [default=false]` to mark this tablet is a MemTablet, and latter add new fields if needed.

One potential issue is there already exists a term `in_memory` in TabletSchema when creating tablet, which means "keep all pages in page caches at best", causes confusion. 

Another option is to add an `enum StorageEngine`, with types of MEMORY and DISK.

### PR DESCRIPTION
**Pull request by decster**
This CL add TabletType to TabletMeta and PartitionInfo, it also adds a create table property "tablet_type" : "disk/memory" for user to specify tablet type. This is a temporary entry point for testing only, it may change in the future, so this field does not persist to FE PartitionInfo(need to upgrade meta version).

Resolves #3442

### CODE REVIEW

File: be/src/olap/base_tablet.h [hunk 1]
```diff
@@ -84,6 +85,10 @@ inline const TabletMetaSharedPtr BaseTablet::tablet_meta() {
     return _tablet_meta;
 }
 
+inline bool BaseTablet::is_memory() const {
+    return _tablet_meta->tablet_type() == TabletTypePB::TABLET_TYPE_MEMORY;
```
Comment (chaoyli) [inline be/src/olap/base_tablet.h:89]: It may be not set default.

File: be/src/olap/tablet_meta.cpp [hunk 2]
```diff
@@ -101,6 +89,8 @@ TabletMeta::TabletMeta(int64_t table_id, int64_t partition_id,
     tablet_meta_pb.set_cumulative_layer_point(-1);
     tablet_meta_pb.set_tablet_state(PB_RUNNING);
     *(tablet_meta_pb.mutable_tablet_uid()) = tablet_uid.to_proto();
+    tablet_meta_pb.set_tablet_type(tabletType == TTabletType::TABLET_TYPE_MEMORY ?
```
Comment (chaoyli) [inline be/src/olap/tablet_meta.cpp:92]: Normal upgrade procedure is upgrading BE process proceeding of upgrading FE.
When upgrading BE, tabletType may be not set.

File: fe/src/main/java/org/apache/doris/alter/SchemaChangeJobV2.java [hunk 3]
```diff
@@ -240,7 +240,8 @@ protected void runPendingJob() throws AlterCancelException {
                                     Partition.PARTITION_INIT_VERSION, Partition.PARTITION_INIT_VERSION_HASH,
                                     tbl.getKeysType(), TStorageType.COLUMN, storageMedium,
                                     shadowSchema, bfColumns, bfFpp, countDownLatch, indexes,
-                                    tbl.isInMemory());
+                                    tbl.isInMemory(),
+                                    tbl.getPartitionInfo().getTabletType(partitionId));
                             createReplicaTask.setBaseTablet(partitionIndexTabletMap.get(partitionId, shadowIdxId).get(shadowTabletId), originSchemaHash);
```
Comment (chaoyli) [inline fe/src/main/java/org/apache/doris/alter/SchemaChangeJobV2.java:245]: You should add constraint about ALTER TABLE.
If I have a disk table.
```
ALTER TABLE test SET("tablet_type"="memory");
```
will be encounter error in BE processing.

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/common/util/PropertyAnalyzer.java [hunk 4]
```diff
@@ -233,6 +236,23 @@ public static TStorageType analyzeStorageType(Map<String, String> properties) th
         return tStorageType;
     }
 
+    public static TTabletType analyzeTabletType(Map<String, String> properties) throws AnalysisException {
+        // default is COLUMN
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/common/util/PropertyAnalyzer.java:240]: modify the comment

Comment (decster) [reply fe/src/main/java/org/apache/doris/common/util/PropertyAnalyzer.java]: fixed

File: be/src/olap/tablet_meta.cpp (hunk 2 above)
Comment (decster) [reply be/src/olap/tablet_meta.cpp:93]: fixed

File: be/src/olap/base_tablet.h (hunk 1 above)
Comment (decster) [reply be/src/olap/base_tablet.h:89]: fixed

**Review by chaoyli — APPROVED** [review]

**Pull request merged**